src/librustc/middle/intrinsicck.rs

```diff
@@ -67,20 +67,6 @@ impl<'a, 'gcx, 'tcx> ExprVisitor<'a, 'gcx, 'tcx> {
             if sk_from.same_size(sk_to) {
                 return;
             }
-
-            match (&from.sty, sk_to) {
-                (&ty::TyFnDef(..), SizeSkeleton::Known(size_to))
-                        if size_to == Pointer.size(&self.infcx.tcx.data_layout) => {
-                    // FIXME #19925 Remove this warning after a release cycle.
-                    let msg = format!("`{}` is now zero-sized and has to be cast \
-                                       to a pointer before transmuting to `{}`",
-                                      from, to);
-                    self.infcx.tcx.sess.add_lint(
-                        ::lint::builtin::TRANSMUTE_FROM_FN_ITEM_TYPES, id, span, msg);
-                    return;
-                }
-                _ => {}
```

Comment:
Let's keep this logic, but use it to give a helper error message and probably a unique error code. This will let us move the text from the tracking issue into a `--explain` that covers some of the key concepts here (i.e., that fn items have a unique zero-sized type). But here in the short form we should give a suggestion for how to rewrite. Something like "try `x as fn(_, _) -> _" or whatever is the most concise form we can come up with. 

I think it's important to get the diagnostics here right, this can be a fairly confusing area for people, and I think it'll come up in FFI and other things that people do early on in Rust.